File: engines/engine_hedge.py
from typing import TYPE_CHECKING

from engines.engine_event import Event, EventEngine
from engines.engine_log import INFO
from strategies.template import OptionStrategyTemplate
from utilities.constant import Direction, OrderType
from utilities.event import EVENT_LOG, EVENT_TIMER
from utilities.object import ContractData, LogData
from utilities.portfolio import UnderlyingData

APP_NAME = "Hedge"

if TYPE_CHECKING:
    from engines.engine_strategy import OptionStrategyEngine


class HedgeConfig:
    """Configuration for strategy hedging"""

    def __init__(self, strategy: OptionStrategyTemplate, timer_trigger: int = 5, delta_target: int = 0, delta_range: int = 0):
        self.strategy = strategy
        self.strategy_name = strategy.strategy_name
        self.timer_trigger = timer_trigger
        self.delta_target = delta_target
        self.delta_range = delta_range


class HedgeEngine:
    """Centralized hedge engine that manages hedging for multiple strategies"""

    def __init__(self, option_strategy_engine: "OptionStrategyEngine") -> None:
        self.option_strategy_engine: "OptionStrategyEngine" = option_strategy_engine
        self.event_engine: EventEngine = option_strategy_engine.event_engine

        # Strategy registration system
        self.registered_strategies: dict[str, HedgeConfig] = {}
        self.timer_count: int = 0
        self.timer_trigger: int = 5

        self.register_event()

    def register_event(self) -> None:
        self.event_engine.register(EVENT_TIMER, self.process_timer_event)

    def register_strategy(
        self, strategy: OptionStrategyTemplate, timer_trigger: int = 5, delta_target: int = 0, delta_range: int = 0
    ) -> None:
        # Validate strategy has underlying symbol
        if not strategy.underlying or not strategy.underlying.symbol:
            self.write_log(f"Cannot register strategy {strategy.strategy_name}: no underlying symbol", INFO)
            return

        config = HedgeConfig(strategy=strategy, timer_trigger=timer_trigger, delta_target=delta_target, delta_range=delta_range)

        self.registered_strategies[strategy.strategy_name] = config
        self.write_log(f"Strategy {strategy.strategy_name} registered for hedging", INFO)

    def unregister_strategy(self, strategy_name: str) -> None:
        if strategy_name in self.registered_strategies:
            del self.registered_strategies[strategy_name]
            self.write_log(f"Strategy {strategy_name} unregistered from hedging", INFO)

    def write_log(self, msg: str, level: int = INFO) -> None:
        # Write log to event system
        log = LogData(msg=msg, level=level, gateway_name=APP_NAME)
        event = Event(EVENT_LOG, log)
        self.event_engine.put(event)

    def process_timer_event(self, event: Event) -> None:
        self.timer_count += 1
        if self.timer_count < self.timer_trigger:
            return
        self.timer_count = 0

        for strategy_name, config in self.registered_strategies.items():
            self.run_strategy_hedging(strategy_name, config)

    def run_strategy_hedging(self, strategy_name: str, config: HedgeConfig) -> None:
        if not self.check_strategy_orders_finished(strategy_name):
            self.cancel_strategy_orders(strategy_name)
            return

        plan = self.compute_hedge_plan(strategy_name, config)
        if not plan:
            return

        symbol, direction, available, order_volume = plan
        self.execute_hedge_orders(strategy_name, symbol, direction, available, order_volume)

    def compute_hedge_plan(self, strategy_name: str, config: HedgeConfig) -> tuple[str, Direction, float, float] | None:
        strategy = self.option_strategy_engine.get_strategy(strategy_name)
        if not strategy:
            return None

        holding = strategy.holding

        total_delta = holding.summary.delta
        delta_max = config.delta_target + config.delta_range
        delta_min = config.delta_target - config.delta_range
        if delta_min <= total_delta <= delta_max:
            return None

        delta_to_hedge = config.delta_target - total_delta
        portfolio = strategy.portfolio
        if portfolio is None or not portfolio.underlying:
            return None

        underlying: UnderlyingData = portfolio.underlying
        hedge_volume = delta_to_hedge / underlying.theo_delta
        symbol = underlying.symbol

        contract: ContractData | None = self.option_strategy_engine.get_contract(symbol)
        if not contract:
            return None
        if abs(hedge_volume) < 1:
            return None

        # Determine direction and available close quantity from strategy holding's underlying
        qty = holding.underlyingPosition.quantity
        if hedge_volume > 0:
            direction = Direction.LONG
            available = abs(qty) if qty < 0 else 0
        else:
            direction = Direction.SHORT
            available = qty if qty > 0 else 0

        return symbol, direction, float(available), float(abs(hedge_volume))

    def execute_hedge_orders(self, strategy_name: str, symbol: str, direction: Direction, available: float, order_volume: float) -> None:
        strategy = self.option_strategy_engine.get_strategy(strategy_name)
        if not strategy:
            return

        remaining = order_volume

        # First CLOSE existing position if available
        if available > 0:
            close_vol = min(available, order_volume)
            self.submit_hedge_order(strategy, symbol, direction, close_vol)
            remaining -= close_vol

        # Then OPEN new hedge if needed
        if remaining > 0:
            self.submit_hedge_order(strategy, symbol, direction, remaining)

    def submit_hedge_order(self, strategy: OptionStrategyTemplate, symbol: str, direction: Direction, volume: float) -> None:
        # Use strategy's underlying order helper methods
        strategy.underlying_order(
            direction=direction, price=0.0, volume=volume, order_type=OrderType.MARKET, reference=f"Hedge_{strategy.strategy_name}"
        )

        self.write_log(f"Hedge sending order: dir={direction.name}, vol={volume}, symbol={symbol}", INFO)

    def check_strategy_orders_finished(self, strategy_name: str) -> bool:
        # Check if strategy has any hedge orders (orders with APP_NAME in reference)
        active_orderids = self.option_strategy_engine.strategy_active_orders.get(strategy_name, set())
        hedge_orderids: set[str] = set()

        for orderid in active_orderids:
            order = self.option_strategy_engine.get_order(orderid)
            if order and order.reference and APP_NAME in order.reference:
                hedge_orderids.add(orderid)

        return len(hedge_orderids) == 0

    def cancel_strategy_orders(self, strategy_name: str) -> None:
        # Use strategy engine's active orders map
        active_orderids = self.option_strategy_engine.strategy_active_orders.get(strategy_name, set())
        hedge_orderids: list[str] = []

        for orderid in active_orderids:
            existing_order = self.option_strategy_engine.get_order(orderid)
            if existing_order and existing_order.reference and APP_NAME in existing_order.reference:
                hedge_orderids.append(orderid)

        # Get strategy instance to pass to cancel_order
        strategy = self.option_strategy_engine.get_strategy(strategy_name)
        if not strategy:
            return

        for orderid in hedge_orderids:
            self.option_strategy_engine.cancel_order(strategy, orderid)

File: Otrader/core/engine_hedge.cpp
/**
 * Shared HedgeEngine (from engines/engine_hedge.py).
 * Same logic: timer, run_strategy_hedging, compute_hedge_plan, execute_hedge_orders, check/cancel
 * orders.
 */

#include "engine_hedge.hpp"

#include <cmath>
#include <math.h>
#include <utility>

namespace engines {

static constexpr const char* APP_NAME = "Hedge";

void HedgeEngine::register_strategy(const std::string& strategy_name, int timer_trigger,
                                    int delta_target, int delta_range) {
    HedgeConfig config;
    config.strategy_name = strategy_name;
    config.timer_trigger = timer_trigger;
    config.delta_target = delta_target;
    config.delta_range = delta_range;
    registered_strategies_[strategy_name] = config;
}

void HedgeEngine::unregister_strategy(const std::string& strategy_name) {
    registered_strategies_.erase(strategy_name);
}

void HedgeEngine::process_hedging(const std::string& strategy_name, const HedgeParams& params,
                                  std::vector<utilities::OrderRequest>* out_orders,
                                  std::vector<utilities::CancelRequest>* out_cancels) {
    if ((out_orders == nullptr) && (out_cancels == nullptr)) {
        return;
    }
    auto it = registered_strategies_.find(strategy_name);
    if (it == registered_strategies_.end()) {
        return;
    }
    run_strategy_hedging_with_params(strategy_name, it->second, params, out_orders, out_cancels);
}

namespace {

inline std::string default_strategy_to_portfolio_name(const std::string& strategy_name) {
    std::string portfolio_name = strategy_name;
    const size_t p = strategy_name.find('_');
    if (p != std::string::npos && p + 1 < strategy_name.size()) {
        portfolio_name = strategy_name.substr(p + 1);
    }
    return portfolio_name;
}

} // namespace

void HedgeEngine::process_timer_event() {
    if (!api_.system.get_strategy_names_ref || !api_.portfolio.get_portfolio ||
        !api_.portfolio.get_holding || !api_.portfolio.get_contract ||
        !api_.execution.get_strategy_active_orders || !api_.execution.get_order ||
        !api_.execution.send_order || !api_.execution.cancel_order) {
        return;
    }

    const auto& strategy_names = api_.system.get_strategy_names_ref();
    if (strategy_names.empty()) {
        return;
    }

    std::unordered_map<std::string, utilities::PortfolioData*> portfolio_cache;
    portfolio_cache.reserve(strategy_names.size());

    std::vector<utilities::OrderRequest> orders;
    std::vector<utilities::CancelRequest> cancels;

    for (const std::string& strategy_name : strategy_names) {
        const std::string portfolio_name = default_strategy_to_portfolio_name(strategy_name);
        utilities::PortfolioData* portfolio_ptr = nullptr;
        if (auto it = portfolio_cache.find(portfolio_name); it != portfolio_cache.end()) {
            portfolio_ptr = it->second;
        } else {
            portfolio_ptr = api_.portfolio.get_portfolio(portfolio_name);
            portfolio_cache.emplace(portfolio_name, portfolio_ptr);
        }

        engines::HedgeParams params;
        params.portfolio = portfolio_ptr;
        params.holding = api_.portfolio.get_holding(strategy_name);
        params.get_contract = api_.portfolio.get_contract;
        params.get_strategy_active_orders = api_.execution.get_strategy_active_orders;
        params.get_order = api_.execution.get_order;

        orders.clear();
        cancels.clear();
        process_hedging(strategy_name, params, &orders, &cancels);

        for (const auto& o : orders) {
            (void)api_.execution.send_order(strategy_name, o);
        }
        for (const auto& c : cancels) {
            api_.execution.cancel_order(c);
        }
        // Per-order string logs are emitted inside submit_hedge_order via BaseEngine::write_log.
    }
}

void HedgeEngine::run_strategy_hedging_with_params(
    const std::string& strategy_name, HedgeConfig& config, const HedgeParams& params,
    std::vector<utilities::OrderRequest>* out_orders,
    std::vector<utilities::CancelRequest>* out_cancels) {
    if (!check_strategy_orders_finished(strategy_name, params)) {
        cancel_strategy_orders(strategy_name, params, out_cancels);
        return;
    }
    auto plan = compute_hedge_plan(strategy_name, config, params);
    if (!plan.has_value()) {
        return;
    }
    auto [symbol, direction, available, order_volume] = plan.value();
    execute_hedge_orders(strategy_name, symbol, direction, available, order_volume, params,
                         out_orders);
}

auto HedgeEngine::compute_hedge_plan(const std::string& strategy_name, HedgeConfig& config,
                                     const HedgeParams& params)
    -> std::optional<std::tuple<std::string, utilities::Direction, double, double>> {
    (void)config;
    if ((params.holding == nullptr) || (params.portfolio == nullptr) ||
        !params.portfolio->underlying) {
        return std::nullopt;
    }

    double total_delta = params.holding->summary.delta;
    double delta_max = config.delta_target + config.delta_range;
    double delta_min = config.delta_target - config.delta_range;
    if (total_delta >= delta_min && total_delta <= delta_max) {
        return std::nullopt;
    }

    double delta_to_hedge = config.delta_target - total_delta;
    utilities::UnderlyingData& underlying = *params.portfolio->underlying;
    double hedge_volume =
        delta_to_hedge / (underlying.theo_delta != 0 ? underlying.theo_delta : 1.0);
    std::string symbol = underlying.symbol;

    if (!params.get_contract) {
        return std::nullopt;
    }
    const utilities::ContractData* contract = params.get_contract(symbol);
    if ((contract == nullptr) || std::abs(hedge_volume) < 1) {
        return std::nullopt;
    }

    int qty = params.holding->underlyingPosition.quantity;
    utilities::Direction direction =
        (hedge_volume > 0) ? utilities::Direction::LONG : utilities::Direction::SHORT;
    double available = (hedge_volume > 0) ? (qty < 0 ? std::abs(qty) : 0.0)
                                          : (qty > 0 ? static_cast<double>(qty) : 0.0);
    return std::make_tuple(symbol, direction, available, std::abs(hedge_volume));
}

void HedgeEngine::execute_hedge_orders(const std::string& strategy_name, const std::string& symbol,
                                       utilities::Direction direction, double available,
                                       double order_volume, const HedgeParams& params,
                                       std::vector<utilities::OrderRequest>* out_orders) {
    if (out_orders == nullptr) {
        return;
    }
    double remaining = order_volume;
    if (available > 0) {
        double close_vol = std::min(available, order_volume);
        submit_hedge_order(strategy_name, symbol, direction, close_vol, params, out_orders);
        remaining -= close_vol;
    }
    if (remaining > 0) {
        submit_hedge_order(strategy_name, symbol, direction, remaining, params, out_orders);
    }
}

void HedgeEngine::submit_hedge_order(const std::string& strategy_name, const std::string& symbol,
                                     utilities::Direction direction, double volume,
                                     const HedgeParams& params,
                                     std::vector<utilities::OrderRequest>* out_orders) {
    if (!params.get_contract) {
        return;
    }
    const utilities::ContractData* contract = params.get_contract(symbol);
    if (contract == nullptr) {
        return;
    }
    utilities::OrderRequest req;
    req.symbol = contract->symbol;
    req.exchange = contract->exchange;
    req.direction = direction;
    req.type = utilities::OrderType::MARKET;
    req.volume = volume;
    req.price = 0.0;
    req.reference = std::string("Hedge_") + strategy_name;
    req.trading_class = contract->trading_class;
    if (out_orders != nullptr) {
        out_orders->push_back(req);
    }

    // Unified core logging path: string-based, via BaseEngine.
    write_log(std::string("Hedge sending order: dir=") +
                  (direction == utilities::Direction::LONG ? "LONG" : "SHORT") +
                  ", vol=" + std::to_string(volume) + ", symbol=" + symbol,
              0, APP_NAME);
}

auto HedgeEngine::check_strategy_orders_finished(const std::string& strategy_name,
                                                 const HedgeParams& params) -> bool {
    if (!params.get_strategy_active_orders || !params.get_order) {
        return true;
    }
    const auto& active_map = params.get_strategy_active_orders();
    auto it = active_map.find(strategy_name);
    if (it == active_map.end()) {
        return true;
    }
    for (const auto& orderid : it->second) {
        utilities::OrderData* order = params.get_order(orderid);
        if ((order != nullptr) && order->reference.find(APP_NAME) != std::string::npos) {
            return false;
        }
    }
    return true;
}

void HedgeEngine::cancel_strategy_orders(const std::string& strategy_name,
                                         const HedgeParams& params,
                                         std::vector<utilities::CancelRequest>* out_cancels) {
    if ((out_cancels == nullptr) || !params.get_strategy_active_orders || !params.get_order) {
        return;
    }
    const auto& active_map = params.get_strategy_active_orders();
    auto it = active_map.find(strategy_name);
    if (it == active_map.end()) {
        return;
    }
    for (const auto& orderid : it->second) {
        utilities::OrderData* order = params.get_order(orderid);
        if ((order != nullptr) && order->reference.find(APP_NAME) != std::string::npos) {
            out_cancels->push_back(order->create_cancel_request());
        }
    }
}

} // namespace engines
